src/celeritas/user/ActionDiagnostic.cc

```diff
@@ -0,0 +1,250 @@
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/user/ActionDiagnostic.cc
+//---------------------------------------------------------------------------//
+#include "ActionDiagnostic.hh"
+
+#include "corecel/Assert.hh"
+#include "corecel/data/CollectionAlgorithms.hh"
+#include "corecel/io/JsonPimpl.hh"
+#include "celeritas/global/ActionRegistry.hh"
+#include "celeritas/phys/ParticleParams.hh"
+
+#include "detail/ActionDiagnosticImpl.hh"
+
+#if CELERITAS_USE_JSON
+#    include <nlohmann/json.hpp>
+
+#    include "corecel/io/LabelIO.json.hh"
+#endif
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+//! Helper function for accumulating data from host and device for all streams
+struct SumCounts
+{
+    std::vector<size_type>* result;
+    std::vector<size_type> temp_host;
+
+    explicit SumCounts(std::vector<size_type>* r) : result(r)
+    {
+        CELER_EXPECT(result);
+    }
+
+    // Transfer host data
+    void operator()(ActionDiagnosticStateData<Ownership::reference,
+                                              MemSpace::host> const& state)
+    {
+        using BinId = ItemId<size_type>;
+
+        CELER_EXPECT(state.counts.size() == result->size());
+        for (auto bin : range(state.counts.size()))
+        {
+            (*result)[bin] += state.counts[BinId{bin}];
+        }
+    }
+
+    // Transfer device data
+    void operator()(ActionDiagnosticStateData<Ownership::reference,
+                                              MemSpace::device> const& state)
+    {
+        CELER_EXPECT(state.counts.size() == result->size());
+
+        if (temp_host.empty())
+        {
+            temp_host.resize(result->size());
+        }
+        copy_to_host(state.counts, make_span(temp_host));
+
+        for (auto bin : range(state.counts.size()))
+        {
+            (*result)[bin] += temp_host[bin];
+        }
+    }
+};
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with action registry and particle data.
+ *
+ * The total number of registered actions is needed in the diagnostic params
+ * data, so the construction of the stream storage is deferred until all
+ * actions have been registered.
+ */
+ActionDiagnostic::ActionDiagnostic(ActionId id,
+                                   SPConstActionRegistry action_reg,
+                                   SPConstParticle particle,
+                                   size_type num_streams)
+    : id_(id)
+    , action_reg_(action_reg)
+    , particle_(particle)
+    , num_streams_(num_streams)
+    , store_(std::make_unique<StoreT>())
```

Comment:
I think it's unnecessary to make this a `unique_ptr` since you use the `StreamStore::operator bool` and only seem to allocate the store here.

Reply:
Since the `exacute()` method is const I wasn't sure how to build the store in there (since we have to wait to build it until all the actions are registered).

Reply:
Oh true. I think making it mutable is acceptable, but so is this way.